core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java

```diff
@@ -776,22 +769,77 @@ protected CacheTopology restoreCacheTopology(ScopedPersistentState state) {
          extraneousMembers.removeAll(persistedCH.getMembers());
          throw CLUSTER.extraneousMembersJoinRestoredCache(extraneousMembers, cacheName);
       }
-      CacheTopology initialTopology = new CacheTopology(initialTopologyId, INITIAL_REBALANCE_ID, persistedCH, null,
+      topologyRestored = true;
+      int topologyId = currentTopology == null ? initialTopologyId : currentTopology.getTopologyId() + 1;
+      CacheTopology initialTopology = new CacheTopology(topologyId, INITIAL_REBALANCE_ID, topologyRestored, persistedCH, null,
             CacheTopology.Phase.NO_REBALANCE, persistedCH.getMembers(), persistentUUIDManager.mapAddresses(persistedCH.getMembers()));
-      setCurrentTopology(initialTopology);
-      setStableTopology(initialTopology);
+      return cacheTopologyCreated(initialTopology);
+   }
+
+   @GuardedBy("this")
+   private CacheTopology cacheTopologyCreated(CacheTopology topology) {
+      setCurrentTopology(topology);
+      setStableTopology(topology);
       rebalancingEnabled = true;
       availabilityMode = AvailabilityMode.AVAILABLE;
-      return initialTopology;
+      return topology;
+   }
+
+   @GuardedBy("this")
+   private boolean restoreTopologyFromState() {
+      assert persistentState.isPresent() : "Persistent state not available";
+      CacheTopology topology = restoreCacheTopology(persistentState.get());
+      if (topology != null) {
+         restoreTopologyFromState(topology);
+         return true;
+      }
+      return false;
    }
 
+   @GuardedBy("this")
+   private void restoreTopologyFromState(CacheTopology topology) {
+      // Change our status
+      status = ComponentStatus.RUNNING;
+      CLUSTER.updatingTopology(cacheName, topology, availabilityMode);
+      eventLogger.info(EventLogCategory.CLUSTER, MESSAGES.cacheMembersUpdated(
+          topology.getMembers(), topology.getTopologyId()));
+      clusterTopologyManager.broadcastTopologyUpdate(cacheName, topology, availabilityMode);
+      clusterTopologyManager.broadcastStableTopologyUpdate(cacheName, topology);
+   }
+
+   public synchronized boolean setCurrentTopologyAsStable(boolean force) {
+      if (currentTopology != null) return false;
+
+      if (persistentState.isPresent()) {
+         List<Address> members = getExpectedMembers();
+         ConsistentHash pastConsistentHash = joinInfo.getConsistentHashFactory()
+               .fromPersistentState(persistentState.get());
+         int missing = pastConsistentHash.getMembers().size() - members.size();
+         int owners = joinInfo.getNumOwners();
+         if (!force && missing >= owners) {
+            throw log.missingTooManyMembers(cacheName, owners, missing, pastConsistentHash.getMembers().size());
+         }
+
+         topologyRestored = missing < owners;
```

Comment:
We could optimize something for REPL, but more bookkeeping is necessary. The first node to join would have all the data, but we don't have the same guarantees for the nodes that join later, as they could resurrect entries.

Reply:
> We could optimize something for REPL, but more bookkeeping is necessary. The first node to join would have all the data, but we don't have the same guarantees for the nodes that join later, as they could resurrect entries.

As we discussed, I really think we have to have purge on startup for any node that joins after the cluster was forced to be available.

Reply:
Yeah, we have this behavior. Summarizing:

* If the cache is replicated, the current nodes do not clear the data. Nodes that join later will purge.
* If the cache is distributed:
* * And missing more or equal to # of owners, the data is purged;
* * Is missing less than # of owners, the current members keep the data, and nodes that join later will purge.

And for all cases, if missing more or equal to the number of owners, the user needs to pass the `-f` to force the topology. We can document this flag as dangerous and leads to purging the state when too many nodes are missing.

Reply:
> And missing more or equal to # of owners, the data is purged;
Can you clarify when this would happen? I wouldn't expect the nodes that were marked as available would purge, just nodes that join after the cluster was made available.

Reply:
For example, in a three nodes cluster with numOwners=2, shut down everything and restart just one node. The single node will reject operations in the cache waiting for the other missing members. Now, with the CLI/REST command, this single node can be forced to operate. Since it is missing two members, which is >= than numOwners, the user must strictly provide the flag to force restart. In this case, if the cache is not replicated, it will purge the data.

Now that I am typing all this, I am just realizing, should we allow this single node to operate with the data it has? The members that join later will be purged in either way.

Reply:
> Now that I am typing all this, I am just realizing, should we allow this single node to operate with the data it has? The members that join later will be purged in either way.

Exactly my thought. We should retain as much data as we can. Only the new node data has to be tossed out as it is stale and we don't know if it is valid or not.